crawl-ref/source/cloud.cc

```diff
@@ -1814,3 +1816,94 @@ void end_still_winds()
     env.level_state &= ~LSTATE_STILL_WINDS;
     mpr("The air resumes its normal movements.");
 }
+
+
+void manage_cloud_generators() {
+    for (CloudGenerator &generator : env.cloud_generators)
+        generator.ran_this_turn = 0;
+    for (CloudGenerator &generator : env.cloud_generators)
+        generator.manage();
+}
+
+void CloudGenerator::manage()
+{
+    if (ran_this_turn == 0) {
+        buildup_turns += you.time_taken;
+        delay -= you.time_taken;
+        if (loc.distance_from(you.pos()) <= 10) {
+            if (special == special_fog_type::normal) {
+
+                int runs = 1;
+                if (type == CLOUD_SALT)
+                    runs = 20;
+
+                while (delay <= 0)
+                {
+                    run(runs);
+                }
+            }
+            if (special == special_fog_type::chained) {
+                buildup_turns += you.time_taken;
+                if (slave_references.empty() && slaves.size() != 0) {
+                    for (CloudGenerator &generator : env.cloud_generators)
+                        for (int i = 0; i < slaves.size(); i++)
+                            if (generator.marker_id == slaves[i] &&
+                                generator.marker_id != marker_id)
+                                slave_references.push_back(&generator);
+                }
+                if (!slave_references.empty()) {
```

Comment:
Uses the pointers from the previous loop to run the chained generator group. Not 100% I did this properly.